<hr style='height:4px; background-color:black; border:none; margin-top:20px; margin-bottom:20px;'>

# **Name:** Tim Hollis  
# **Course:** DSC540 - Data Preparation  
# **Date:**  02/15/2026
# **Assignment:** Week 7 & 8 Exercises 


## **Assignment methods and datasets by chapter:**

**Chapter 7**
- Filter out missing data
- Fill in missing data
- Remove duplicates
- Transform data using either mapping or a function
- Replace values
- Discretization and Binning
- Manipulate Strings

***Dataset: MetObjects.csv***

**Chapter 8**
- Create hierarchical index
- Combine and Merge Datasets (you will have to either create a new dataset from your existing data or create a relationship between the data I have provided)
- Reshape
- Pivot the data

***Datasets: BOING-BOING-CANDY-HIERARCHY-2016-SURVEY-Responses.xlsx/ CANDYDATA.xlsx/ CANDY-HIERARCHY-2015-SURVEY-Responses.xlsx/ candyhierarchy2017.xlsx/ candyhierarchysurvey2017.pdf***

**Chapter 10**
- Grouping with Dicts/Series
- Grouping with Functions
- Grouping with Index Levels
- Split/Apply/Combine
- Cross Tabs

***Datasets: BOING-BOING-CANDY-HIERARCHY-2016-SURVEY-Responses.xlsx/ CANDYDATA.xlsx/ CANDY-HIERARCHY-2015-SURVEY-Responses.xlsx/ candyhierarchy2017.xlsx***

**Chapter 11**
- Convert between string and date time
- Generate date range
- Frequencies and date offsets
- Convert timestamps to periods and back
- Period Frequency conversions

***Dataset: MetObjects.csv***

## 📊 Dataset Usage Summary

| Chapter | Methods Demonstrated | Dataset(s) Used |
|--------|-----------------------|-----------------|
| **7** | Cleaning, transforming, binning, replacing, string ops | **MetObjects.csv** |
| **8** | Merge, combine, reshape, pivot, hierarchical index | **Candy Excel datasets** |
| **10** | Grouping, split–apply–combine, crosstabs | **Candy Excel datasets** |
| **11** | Datetime conversion, ranges, periods, offsets | **MetObjects.csv** |

---

## 🏛 Attribution Note 

**Source:** 
This notebook uses The Metropolitan Museum of Art’s Open Access dataset `MetObjects.csv`.  
Data is provided under a Creative Commons Zero (CC0) license courtesy of The Metropolitan Museum of Art.

<hr style='height:2px; background-color:black; border:none; margin-top:20px; margin-bottom:20px;'


## **Initial Setup:**

In [14]:
# Loading libraries
import pandas as pd
import csv
import tabula
import pdfplumber

# Show all columns when printing DataFrames
pd.set_option('display.max_columns', None)

# Loading datasets
met = pd.read_csv('MetObjects.csv', low_memory=False)
df_2015 = pd.read_excel('CANDY-HIERARCHY-2015-SURVEY-Responses.xlsx')
df_2016 = pd.read_excel(
    'BOING-BOING-CANDY-HIERARCHY-2016-SURVEY-Responses.xlsx')
df_2017 = pd.read_excel('candyhierarchy2017.xlsx')
df_candydata = pd.read_excel('CANDYDATA.xlsx')
datasets = {
    '2015 Survey': df_2015,
    '2016 Survey': df_2016,
    '2017 Survey': df_2017,
    'Candy Data': df_candydata
}

# PDF extraction with fallback due to errors with Tabula
try:
    pdf_tables = tabula.read_pdf(
        'candyhierarchysurvey2017.pdf',
        pages='all',
        multiple_tables=True,
        encoding='latin1'
    )

# Previewing datasets
    print('\nPDF Table 1 — First 10 Rows')
    display(pdf_tables[0].head(10))

except Exception as e:
    print('\nTabula failed. Using pdfplumber...\n')
    with pdfplumber.open('candyhierarchysurvey2017.pdf') as pdf:
        page = pdf.pages[0]
        text = page.extract_text()
        print(text[:2000])

print()  # adding space
print('MetObjects.csv — First 10 Rows')
display(met.head(10))

print('\nMetObjects.csv — Info')
met.info()

for name, data in datasets.items():
    print(f'\n{name} — First 10 Rows')
    display(data.head(10))

    print(f'\n{name} — Info')
    data.info()

C:\Users\slimt\anaconda3\envs\DSC530\Lib\site-packages\openpyxl\worksheet\_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)
C:\Users\slimt\anaconda3\envs\DSC530\Lib\site-packages\openpyxl\worksheet\_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)



Tabula failed. Using pdfplumber...

10/25/2017 CandyHierarchy2017 - 0%
CandyHierarchy2017
0%
CANDY HIERARCHY 2017 SURVEY (~10 minutes)
This is the official survey form for the 2017 Candy Hierarchy. All data collected is anonymous and will be shared with the public as a raw spreadsheet (xml) file. This
data is not intended for research purposes, but rather for the writing of a satirical/humour creative non­fiction science piece to be published at BoingBoing. Note that
although unlikely (especially given the nature of the questions), there is always the possibility of someone being able to identify you based on your responses ­ as a
result, please feel free to skip questions as you see fit. Note that although the survey will remain open until Halloween, we will likely only use the data collected up to
noon (PST) of October 25th, 2017 for our formal publication of analysis at BoingBoing. See also last year's Candy Hierarchy (Cohen and Ng, 2016) for context.
https://boingboing.net/2016/10

,Object Number,Is Highlight,Is Timeline Work,Is Public Domain,Object ID,Gallery Number,Department,AccessionYear,Object Name,Title,Culture,Period,Dynasty,Reign,Portfolio,Constituent ID,Artist Role,Artist Prefix,Artist Display Name,Artist Display Bio,Artist Suffix,Artist Alpha Sort,Artist Nationality,Artist Begin Date,Artist End Date,Artist Gender,Artist ULAN URL,Artist Wikidata URL,Object Date,Object Begin Date,Object End Date,Medium,Dimensions,Credit Line,Geography Type,City,State,County,Country,Region,Subregion,Locale,Locus,Excavation,River,Classification,Rights and Reproduction,Link Resource,Object Wikidata URL,Metadata Date,Repository,Tags,Tags AAT URL,Tags Wikidata URL
0,1979.486.1,False,False,False,1,NaN,The American Wing,1979,Coin,One-dollar Liberty Head Coin,NaN,NaN,NaN,NaN,NaN,16429,Maker,,James Barton Longacre,"American, Delaware County, Pennsylvania 1794–1...",,"Longacre, James Barton",American,1794,1869,NaN,http://vocab.getty.edu/page/ulan/500011409,https://www.wikidata.org/wiki/Q3806459,1853,1853,1853,Gold,Dimensions unavailable,"Gift of Heinz L. Stoppelmann, 1979",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,http://www.metmuseum.org/art/collection/search/1,NaN,NaN,"Metropolitan Museum of Art, New York, NY",NaN,NaN,NaN
1,1980.264.5,False,False,False,2,NaN,The American Wing,1980,Coin,Ten-dollar Liberty Head Coin,NaN,NaN,NaN,NaN,NaN,107,Maker,,Christian Gobrecht,1785–1844,,"Gobrecht, Christian",American,1785,1844,NaN,http://vocab.getty.edu/page/ulan/500077295,https://www.wikidata.org/wiki/Q5109648,1901,1901,1901,Gold,Dimensions unavailable,"Gift of Heinz L. Stoppelmann, 1980",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,http://www.metmuseum.org/art/collection/search/2,NaN,NaN,"Metropolitan Museum of Art, New York, NY",NaN,NaN,NaN
2,67.265.9,False,False,False,3,NaN,The American Wing,1967,Coin,Two-and-a-Half Dollar Coin,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1909–27,1909,1927,Gold,Diam. 11/16 in. (1.7 cm),"Gift of C. Ruxton Love Jr., 1967",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,http://www.metmuseum.org/art/collection/search/3,NaN,NaN,"Metropolitan Museum of Art, New York, NY",NaN,NaN,NaN
3,67.265.10,False,False,False,4,NaN,The American Wing,1967,Coin,Two-and-a-Half Dollar Coin,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1909–27,1909,1927,Gold,Diam. 11/16 in. (1.7 cm),"Gift of C. Ruxton Love Jr., 1967",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,http://www.metmuseum.org/art/collection/search/4,NaN,NaN,"Metropolitan Museum of Art, New York, NY",NaN,NaN,NaN
4,67.265.11,False,False,False,5,NaN,The American Wing,1967,Coin,Two-and-a-Half Dollar Coin,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1909–27,1909,1927,Gold,Diam. 11/16 in. (1.7 cm),"Gift of C. Ruxton Love Jr., 1967",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,http://www.metmuseum.org/art/collection/search/5,NaN,NaN,"Metropolitan Museum of Art, New York, NY",NaN,NaN,NaN
5,67.265.12,False,False,False,6,NaN,The American Wing,1967,Coin,Two-and-a-Half Dollar Coin,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1909–27,1909,1927,Gold,Diam. 11/16 in. (1.7 cm),"Gift of C. Ruxton Love Jr., 1967",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,http://www.metmuseum.org/art/collection/search/6,NaN,NaN,"Metropolitan Museum of Art, New York, NY",NaN,NaN,NaN
6,67.265.13,False,False,False,7,NaN,The American Wing,1967,Coin,Two-and-a-Half Dollar Coin,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1909–27,1909,1927,Gold,Diam. 11/16 in. (1.7 cm),"Gift of C. Ruxton Love Jr., 1967",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,http://www.metmuseum.org/art/collection/search/7,NaN,NaN,"Metropolitan Museum of Art, New York, NY",Birds|Coins,http://vocab.getty.edu/page/aat/300266506|http...,https://www.wikidata.org/wiki/Q5113|https://ww...
7,67.265.14,False,False,False,8,NaN,The American Wing,1967,Coin,Two


MetObjects.csv — Info
<class 'pandas.DataFrame'>
RangeIndex: 484956 entries, 0 to 484955
Data columns (total 54 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Object Number            484956 non-null  str    
 1   Is Highlight             484956 non-null  bool   
 2   Is Timeline Work         484956 non-null  bool   
 3   Is Public Domain         484956 non-null  bool   
 4   Object ID                484956 non-null  int64  
 5   Gallery Number           49541 non-null   str    
 6   Department               484956 non-null  str    
 7   AccessionYear            481094 non-null  str    
 8   Object Name              482690 non-null  str    
 9   Title                    456153 non-null  str    
 10  Culture                  208190 non-null  str    
 11  Period                   91143 non-null   str    
 12  Dynasty                  23201 non-null   str    
 13  Reign                    11236 non-null   str  

,Timestamp,How old are you?,Are you going actually going trick or treating yourself?,[Butterfinger],[100 Grand Bar],[Anonymous brown globs that come in black and orange wrappers],[Any full-sized candy bar],[Black Jacks],[Bonkers],[Bottle Caps],[Box’o’ Raisins],[Brach products (not including candy corn)],[Bubble Gum],[Cadbury Creme Eggs],[Candy Corn],"[Vials of pure high fructose corn syrup, for main-lining into your vein]",[Candy that is clearly just the stuff given out for free at restaurants],"[Cash, or other forms of legal tender]",[Chiclets],[Caramellos],[Snickers],[Dark Chocolate Hershey],[Dental paraphenalia],[Dots],[Fuzzy Peaches],[Generic Brand Acetaminophen],[Glow sticks],[Broken glow stick],[Goo Goo Clusters],[Good N' Plenty],[Gum from baseball cards],[Gummy Bears straight up],[Creepy Religious comics/Chick Tracts],[Healthy Fruit],[Heath Bar],[Hershey’s Kissables],[Hershey’s Milk Chocolate],[Hugs (actual physical hugs)],[Jolly Rancher (bad flavor)],[Jolly Ranchers (good flavor)],[Kale smoothie],[Kinder Happy Hippo],[Kit Kat],[Hard Candy],[Lapel Pins],[LemonHeads],[Licorice],[Licorice (not black)],[Lindt Truffle],[Lollipops],[Mars],[Mary Janes],[Maynards],[Milk Duds],[LaffyTaffy],[Minibags of chips],[JoyJoy (Mit Iodine)],[Reggie Jackson Bar],[Pixy Stix],[Nerds],[Nestle Crunch],[Now'n'Laters],[Pencils],[Milky Way],[Reese’s Peanut Butter Cups],[Tolberone something or other],[Runts],[Junior Mints],[Senior Mints],[Mint Kisses],[Mint Juleps],[Mint Leaves],[Peanut M&M’s],[Regular M&Ms],[Mint M&Ms],[Ribbon candy],[Rolos],[Skittles],[Smarties (American)],[Smarties (Commonwealth)],[Chick-o-Sticks (we don’t know what that is)],[Spotted Dick],[Starburst],[Swedish Fish],[Sweetums],[Those odd marshmallow circus peanut things],[Three Musketeers],[Peterson Brand Sidewalk Chalk],[Peanut Butter Bars],[Peanut Butter Jars],[Trail Mix],[Twix],[Vicodin],[White Bread],[Whole Wheat anything],[York Peppermint Patties],Please leave any remarks or comments regarding your choices.,Please list any items not included above that give you JOY.,Please list any items not included above that give you DESPAIR.,Guess the number of mints in my hand.,Betty or Veronica?,"Check all that apply: ""I cried tears of sadness at the end of ____________""","""That dress* that went viral early this year - when I first saw it, it was ________""","Fill in the blank: ""Taylor Swift is a force for ___________""",What is your favourite font?,"If you squint really hard, the words ""Intelligent Design"" would look like.","Fill in the blank: ""Imitation is a form of ____________""",Please estimate the degree(s) of separation you have from the following celebrities [JK Rowling],Please estimate the degree(s) of separation you have from the following celebrities [JJ Abrams],Please estimate the degree(s) of separation you have from the following celebrities [Beyoncé],Please estimate the degree(s) of separation you have from the following celebrities [Bieber],Please estimate the degree(s) of separation you have from the following celebrities [Kevin Bacon],Please estimate the degree(s) of separation you have from the following celebrities [Francis Bacon (1561 - 1626)],"[Sea-salt flavored stuff, probably chocolate, since this is the ""it"" flavor of the year]",[Necco Wafers],"Which day do you prefer, Friday or Sunday?",Please estimate the degrees of separation you have from the following folks [Bruce Lee],Please estimate the degrees of separation you have from the following folks [JK Rowling],Please estimate the degrees of separation you have from the following folks [Malala Yousafzai],Please estimate the degrees of separation you have from the following folks [Thom Yorke],Please estimate the degrees of separation you have from the following folks [JJ Abrams],Please estimate the degrees of separation you have from the following folks [Hillary Clinton],Please estimate the degrees of separation you have from the following folks [Donald Trump],Please estimate the degrees of separati


2015 Survey — Info
<class 'pandas.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Columns: 124 entries, Timestamp to Please estimate the degrees of separation you have from the following folks [Beyoncé Knowles]
dtypes: datetime64[us](1), float64(9), object(11), str(103)
memory usage: 5.3+ MB

2016 Survey — First 10 Rows


,Timestamp,Are you going actually going trick or treating yourself?,Your gender:,How old are you?,Which country do you live in?,"Which state, province, county do you live in?",[100 Grand Bar],[Anonymous brown globs that come in black and orange wrappers],[Any full-sized candy bar],[Black Jacks],[Bonkers (the candy)],[Bonkers (the board game)],[Bottle Caps],[Box'o'Raisins],[Broken glow stick],[Butterfinger],[Cadbury Creme Eggs],[Candy Corn],[Candy that is clearly just the stuff given out for free at restaurants],[Caramellos],"[Cash, or other forms of legal tender]",[Chardonnay],[Chick-o-Sticks (we don’t know what that is)],[Chiclets],[Coffee Crisp],[Creepy Religious comics/Chick Tracts],[Dental paraphenalia],[Dots],[Dove Bars],[Fuzzy Peaches],[Generic Brand Acetaminophen],[Glow sticks],[Goo Goo Clusters],[Good N' Plenty],[Gum from baseball cards],[Gummy Bears straight up],[Hard Candy],[Healthy Fruit],[Heath Bar],[Hershey's Dark Chocolate],[Hershey’s Milk Chocolate],[Hershey's Kisses],[Hugs (actual physical hugs)],[Jolly Rancher (bad flavor)],[Jolly Ranchers (good flavor)],[JoyJoy (Mit Iodine!)],[Junior Mints],[Senior Mints],[Kale smoothie],[Kinder Happy Hippo],[Kit Kat],[LaffyTaffy],[LemonHeads],[Licorice (not black)],[Licorice (yes black)],[Lindt Truffle],[Lollipops],[Mars],[Mary Janes],[Maynards],[Mike and Ike],[Milk Duds],[Milky Way],[Regular M&Ms],[Peanut M&M’s],[Blue M&M's],[Red M&M's],[Third Party M&M's],[Minibags of chips],[Mint Kisses],[Mint Juleps],[Mr. Goodbar],[Necco Wafers],[Nerds],[Nestle Crunch],[Now'n'Laters],[Peeps],[Pencils],[Person of Interest Season 3 DVD Box Set (not including Disc 4 with hilarious outtakes)],[Pixy Stix],[Reese’s Peanut Butter Cups],[Reese's Pieces],[Reggie Jackson Bar],[Rolos],[Skittles],[Smarties (American)],[Smarties (Commonwealth)],[Snickers],[Sourpatch Kids (i.e. abominations of nature)],[Spotted Dick],[Starburst],[Sweet Tarts],[Swedish Fish],[Sweetums (a friend to diabetes)],[Tic Tacs],[Those odd marshmallow circus peanut things],[Three Musketeers],[Tolberone something or other],[Trail Mix],[Twix],"[Vials of pure high fructose corn syrup, for main-lining into your vein]",[Vicodin],[Whatchamacallit Bars],[White Bread],[Whole Wheat anything],[York Peppermint Patties],Please list any items not included above that give you JOY.,Please list any items not included above that give you DESPAIR.,"Please leave any witty, snarky or thoughtful remarks or comments regarding your choices.",Guess the number of mints in my hand.,Betty or Veronica?,"""That dress* that went viral a few years back - when I first saw it, it was ________""",What is your favourite font?,Please estimate the degree(s) of separation you have from the following celebrities [JK Rowling],Please estimate the degree(s) of separation you have from the following celebrities [JJ Abrams],Please estimate the degree(s) of separation you have from the following celebrities [Beyoncé],Please estimate the degree(s) of separation you have from the following celebrities [Bieber],Please estimate the degree(s) of separation you have from the following celebrities [Kevin Bacon],Please estimate the degree(s) of separation you have from the following celebrities [Francis Bacon (1561 - 1626)],"Which day do you prefer, Friday or Sunday?","Do you eat apples the correct way, East to West (side to side) or do you eat them like a freak of nature, South to North (bottom to top)?","When you see the above image of the 4 different websites, which one would you most likely check out (please be honest).",[York Peppermint Patties] Ignore
0,2016-10-24 05:09:23.033,No,Male,22,Canada,Ontario,JOY,DESPAIR,JOY,MEH,MEH,MEH,MEH,DESPAIR,DESPAIR,JOY,JOY,DESPAIR,DESPAIR,JOY,JOY,MEH,MEH,MEH,JOY,JOY,DESPAIR,DESPAIR,DESPAIR,JOY,DESPAIR,DESPAIR,DESPAIR,JOY,DESPAIR,JOY,JOY,DESPAIR,DESPAIR,JOY,JOY,JOY,MEH,MEH,JOY,DESPAIR,JOY,JOY,DESPAIR,JOY,JOY,JOY,JOY,JOY,DESPAIR,JOY,JOY,JOY,JOY,JOY,JOY,JOY,JOY,JOY,JOY,JOY,JOY,MEH,JOY,JOY,JOY,JOY,DESPAIR,JOY,DESPAIR,DESPAIR,DESPAIR,DESPAIR,MEH


2016 Survey — Info
<class 'pandas.DataFrame'>
RangeIndex: 1259 entries, 0 to 1258
Columns: 123 entries, Timestamp to  [York Peppermint Patties] Ignore
dtypes: datetime64[us](1), float64(1), object(10), str(111)
memory usage: 1.2+ MB

2017 Survey — First 10 Rows


,Internal ID,Q1: GOING OUT?,Q2: GENDER,Q3: AGE,Q4: COUNTRY,"Q5: STATE, PROVINCE, COUNTY, ETC",Q6 | 100 Grand Bar,Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes),Q6 | Any full-sized candy bar,Q6 | Black Jacks,Q6 | Bonkers (the candy),Q6 | Bonkers (the board game),Q6 | Bottle Caps,Q6 | Box'o'Raisins,Q6 | Broken glow stick,Q6 | Butterfinger,Q6 | Cadbury Creme Eggs,Q6 | Candy Corn,Q6 | Candy that is clearly just the stuff given out for free at restaurants,Q6 | Caramellos,"Q6 | Cash, or other forms of legal tender",Q6 | Chardonnay,Q6 | Chick-o-Sticks (we don’t know what that is),Q6 | Chiclets,Q6 | Coffee Crisp,Q6 | Creepy Religious comics/Chick Tracts,Q6 | Dental paraphenalia,Q6 | Dots,Q6 | Dove Bars,Q6 | Fuzzy Peaches,Q6 | Generic Brand Acetaminophen,Q6 | Glow sticks,Q6 | Goo Goo Clusters,Q6 | Good N' Plenty,Q6 | Gum from baseball cards,Q6 | Gummy Bears straight up,Q6 | Hard Candy,Q6 | Healthy Fruit,Q6 | Heath Bar,Q6 | Hershey's Dark Chocolate,Q6 | Hershey’s Milk Chocolate,Q6 | Hershey's Kisses,Q6 | Hugs (actual physical hugs),Q6 | Jolly Rancher (bad flavor),Q6 | Jolly Ranchers (good flavor),Q6 | JoyJoy (Mit Iodine!),Q6 | Junior Mints,Q6 | Senior Mints,Q6 | Kale smoothie,Q6 | Kinder Happy Hippo,Q6 | Kit Kat,Q6 | LaffyTaffy,Q6 | LemonHeads,Q6 | Licorice (not black),Q6 | Licorice (yes black),Q6 | Lindt Truffle,Q6 | Lollipops,Q6 | Mars,Q6 | Maynards,Q6 | Mike and Ike,Q6 | Milk Duds,Q6 | Milky Way,Q6 | Regular M&Ms,Q6 | Peanut M&M’s,Q6 | Blue M&M's,Q6 | Red M&M's,Q6 | Green Party M&M's,Q6 | Independent M&M's,Q6 | Abstained from M&M'ing.,Q6 | Minibags of chips,Q6 | Mint Kisses,Q6 | Mint Juleps,Q6 | Mr. Goodbar,Q6 | Necco Wafers,Q6 | Nerds,Q6 | Nestle Crunch,Q6 | Now'n'Laters,Q6 | Peeps,Q6 | Pencils,Q6 | Pixy Stix,Q6 | Real Housewives of Orange County Season 9 Blue-Ray,Q6 | Reese’s Peanut Butter Cups,Q6 | Reese's Pieces,Q6 | Reggie Jackson Bar,Q6 | Rolos,Q6 | Sandwich-sized bags filled with BooBerry Crunch,Q6 | Skittles,Q6 | Smarties (American),Q6 | Smarties (Commonwealth),Q6 | Snickers,Q6 | Sourpatch Kids (i.e. abominations of nature),Q6 | Spotted Dick,Q6 | Starburst,Q6 | Sweet Tarts,Q6 | Swedish Fish,Q6 | Sweetums (a friend to diabetes),Q6 | Take 5,Q6 | Tic Tacs,Q6 | Those odd marshmallow circus peanut things,Q6 | Three Musketeers,Q6 | Tolberone something or other,Q6 | Trail Mix,Q6 | Twix,"Q6 | Vials of pure high fructose corn syrup, for main-lining into your vein",Q6 | Vicodin,Q6 | Whatchamacallit Bars,Q6 | White Bread,Q6 | Whole Wheat anything,Q6 | York Peppermint Patties,Q7: JOY OTHER,Q8: DESPAIR OTHER,Q9: OTHER COMMENTS,Q10: DRESS,Unnamed: 113,Q11: DAY,Q12: MEDIA [Daily Dish],Q12: MEDIA [Science],Q12: MEDIA [ESPN],Q12: MEDIA [Yahoo],"Click Coordinates (x, y)"
0,90258773,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,90272821,No,Male,44,USA,NM,MEH,DESPAIR,JOY,MEH,DESPAIR,DESPAIR,DESPAIR,DESPAIR,DESPAIR,DESPAIR,MEH,MEH,DESPAIR,MEH,JOY,MEH,DESPAIR,DESPAIR,DESPAIR,DESPAIR,DESPAIR,MEH,JOY,DESPAIR,DESPAIR,DESPAIR,DESPAIR,MEH,DESPAIR,MEH,MEH,DESPAIR,MEH,JOY,JOY,MEH,DESPAIR,DESPAIR,MEH,DESPAIR,DESPAIR,DESPAIR,DESPAIR,DESPAIR,JOY,DESPAIR,MEH,MEH,JOY,MEH,DESPAIR,DESPAIR,DESPAIR,MEH,MEH,JOY,JOY,MEH,JOY,JOY,JOY,JOY,DESPAIR,DESPAIR,MEH,DESPAIR,DESPAIR,DESPAIR,DESPAIR,JOY,DESPAIR,DESPAIR,DESPAIR,DESPAIR,DESPAIR,JOY,JOY,DESPAIR,JOY,DESPAIR,DESPAIR,DESPAIR,DESPAIR,MEH,DESPAIR,DESPAIR,MEH,DESPAIR,MEH,DESPAIR,DESPAIR,DESPAIR,DESPAIR,JOY,JOY,DESPAIR,JOY,DESPAIR,DESPAIR,DESPAIR,DESPAIR,DESPAIR,DESPAIR,Mounds,NaN,Bottom line is Twix is really the only candy w...,White and gold,NaN,Sunday,Na


2017 Survey — Info
<class 'pandas.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Columns: 120 entries, Internal ID to Click Coordinates (x, y)
dtypes: float64(4), int64(1), object(4), str(111)
memory usage: 2.3+ MB

Candy Data — First 10 Rows


,ITEM,JOY,DESPAIR,NET FEELIES,NET CLOUT,DESPAIR (NEG)
0,York Peppermint Patties,634,78,556.0,1.639118,-78.0
1,Whole Wheat anything,21,419,-398.0,1.012938,-419.0
2,White Bread,15,473,-458.0,1.123440,-473.0
3,Vicodin,323,210,113.0,1.227036,-210.0
4,Twix,770,26,744.0,1.832497,-26.0
5,Trail Mix,81,247,-166.0,0.755099,-247.0
6,Tolberone something or other,627,44,583.0,1.544730,-44.0
7,Three Musketeers,534,40,494.0,1.321424,-40.0
8,Those odd marshmallow circus peanut things,135,545,-410.0,1.565450,-545.0
9,Sweetums,21,92,-71.0,0.260141,-92.0



Candy Data — Info
<class 'pandas.DataFrame'>
RangeIndex: 87 entries, 0 to 86
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ITEM           86 non-null     str    
 1   JOY            87 non-null     int64  
 2   DESPAIR        87 non-null     int64  
 3   NET FEELIES    86 non-null     float64
 4   NET CLOUT      86 non-null     float64
 5   DESPAIR (NEG)  86 non-null     float64
dtypes: float64(3), int64(2), str(1)
memory usage: 4.2 KB


<hr style="border: none; height: 3px; background-color: black;">

# Chapter 7 — Data Cleaning and Transformation (Part 1)

In this first part of Chapter 7, I focus on the foundational cleaning steps using the MetObjects dataset. These operations help identify missing values, remove incomplete rows, fill in gaps with appropriate defaults, and eliminate duplicate records. Together, they establish a clean and reliable starting point for the transformations that follow.

### Steps for this section (1–4)

1. Use `.isna().sum()` to identify which columns contain missing values and understand the overall completeness of the dataset.  
2. Use `.dropna()` to filter out rows that are incomplete based on specific fields that are required for analysis.  
3. Use `.fillna()` to replace missing values with defaults such as `'Unknown'` for text fields or `0` for numeric fields.  
4. Use `.drop_duplicates()` to remove duplicate records based on a key column to ensure each entry is unique.

<hr style="border: none; height: 3px; background-color: black;">

In [15]:
# 1: Identify missing values across all columns
missing_counts = met.isna().sum()
print('Missing values per column:')
print(missing_counts)

# 2: Filter out rows missing key descriptive fields
before_rows = len(met)

# Remove rows missing the artist name
met_no_missing_artist = met.dropna(subset=['Artist Display Name'])
after_artist_rows = len(met_no_missing_artist)

# Remove rows missing both Object Name and Title
met_no_missing_obj_title = met.dropna(
    subset=['Object Name', 'Title'], how='all')
after_obj_title_rows = len(met_no_missing_obj_title)

print(
    '\nRows removed due to missing Artist Display Name:',
    before_rows -
    after_artist_rows)
print('Rows removed due to missing Object Name and Title:',
      before_rows - after_obj_title_rows)

# 3: Fill in missing values with appropriate defaults
met_filled = met.copy()

before_culture_missing = met_filled['Culture'].isna().sum()
before_objectid_missing = met_filled['Object ID'].isna().sum()

# Replace missing Culture with 'Unknown' and missing Object ID with 0
met_filled['Culture'] = met_filled['Culture'].fillna('Unknown')
met_filled['Object ID'] = met_filled['Object ID'].fillna(0)

after_culture_missing = met_filled['Culture'].isna().sum()
after_objectid_missing = met_filled['Object ID'].isna().sum()

print(
    '\nFilled missing Culture values:',
    before_culture_missing -
    after_culture_missing)
print(
    'Filled missing Object ID values:',
    before_objectid_missing -
    after_objectid_missing)

# 4: Remove duplicate records based on Object ID
before_dupes = len(met)
met_no_duplicates = met.drop_duplicates(subset=['Object ID'])
after_dupes = len(met_no_duplicates)

print('\nDuplicate rows removed based on Object ID:', before_dupes - after_dupes)

Missing values per column:
Object Number                   0
Is Highlight                    0
Is Timeline Work                0
Is Public Domain                0
Object ID                       0
Gallery Number             435415
Department                      0
AccessionYear                3862
Object Name                  2266
Title                       28803
Culture                    276766
Period                     393813
Dynasty                    461755
Reign                      473720
Portfolio                  458442
Constituent ID             202443
Artist Role                202443
Artist Prefix              202443
Artist Display Name        202443
Artist Display Bio         204533
Artist Suffix              202491
Artist Alpha Sort          202443
Artist Nationality         202443
Artist Begin Date          202443
Artist End Date            202443
Artist Gender              378474
Artist ULAN URL            257515
Artist Wikidata URL        260754
Object Date          

<div style="border: 2px solid black; padding: 10px;">

### Summary (Part 1)

After reviewing the MetObjects dataset, several key data‑quality issues became clear. Many fields contain missing values, especially those related to artist information, geography, and descriptive metadata. The steps in this section focused on identifying these gaps, removing incomplete rows, filling in missing values where appropriate, and checking for duplicate records.

**Missing Values Overview**  
A number of columns contain substantial missing data. For example, *Culture* has 276,766 missing entries, *Period* has 393,813 missing entries, and *Artist Display Name* has 202,443 missing entries. Other fields such as *Gallery Number*, *City*, *Region*, and *Metadata Date* also contain large gaps. These patterns confirmed the need for targeted filtering and filling.

**Rows Removed Due to Missing Key Fields**  
- Rows missing *Artist Display Name*: **202,443** removed  
- Rows missing both *Object Name* and *Title*: **53** removed  

These fields are essential for identifying and classifying objects, so removing incomplete rows helps maintain consistency and reliability in later analysis.

**Missing Values Filled**  
- Filled missing values in *Culture*: **276,766** entries updated to `'Unknown'`  
- Filled missing values in *Object ID*: **0** entries updated (no missing values were present)  

Filling these values prevents issues during grouping, filtering, and descriptive analysis, especially when working with categorical fields.

**Duplicate Records Removed**  
- Duplicate rows based on *Object ID*: **0** duplicates found  

This confirms that the dataset does not contain repeated object identifiers, and no further action was required for deduplication.

<hr style="border: none; height: 3px; background-color: black;">

# Chapter 7 — Data Cleaning and Transformation (Part 2)

This part of Chapter 7 focuses on transforming values, replacing specific entries, creating bins for numeric fields, and working with text fields. These operations help standardize the dataset, extract meaningful features, and prepare the data for grouping, visualization, and later analysis.

### Steps for this section

1. Use `.map()` and `.apply()` to transform values or apply custom functions to selected columns.  
2. Use `.replace()` to update specific values or apply a dictionary of replacements for consistency.  
3. Use `pd.cut()` and `pd.qcut()` to create equal‑width or quantile‑based bins for numeric fields.  
4. Use string methods such as `.str.upper()`, `.str.contains()`, and `.str.split()` to clean and extract information from text fields.

<hr style="border: none; height: 3px; background-color: black;">

In [16]:
# 1. Transform values using mapping and a custom function

# Map the boolean public-domain flag to human-readable labels
pub_map = {True: 'Public', False: 'Restricted'}
met['PublicStatus'] = met['Is Public Domain'].map(pub_map)

# Extract numeric year values from the AccessionYear field


def extract_year(val):
    try:
        return int(val)
    except BaseException:
        return None


met['AccessionYear_Extracted'] = met['AccessionYear'].apply(extract_year)

mapped_count = met['PublicStatus'].notna().sum()
year_extracted_count = met['AccessionYear_Extracted'].notna().sum()

print('Values mapped for PublicStatus:', mapped_count)
print('AccessionYear values successfully extracted:', year_extracted_count)

# 2. Replace values for consistency across categories
met_replaced = met.copy()

# Replace 'Unknown' culture values with a clearer label
before_unknown = (met_replaced['Culture'] == 'Unknown').sum()
met_replaced['Culture'] = met_replaced['Culture'].replace(
    'Unknown', 'Not Specified')
after_unknown = (met_replaced['Culture'] == 'Not Specified').sum()

# Replace selected department names using a mapping dictionary
dept_map = {'European Paintings': 'Euro Paintings', 'Asian Art': 'Asia Dept'}
before_dept = met_replaced['Department'].isin(dept_map.keys()).sum()
met_replaced['Department'] = met_replaced['Department'].replace(dept_map)
after_dept = met_replaced['Department'].isin(dept_map.values()).sum()

print('\nCulture values replaced:', after_unknown)
print('Department values replaced:', after_dept)

# 3. Create bins for numeric fields using equal-width and quantile-based
# methods
met_bins = met.copy()

met_bins['ID_width_bins'] = pd.cut(met_bins['Object ID'], bins=4)
met_bins['ID_quantile_bins'] = pd.qcut(
    met_bins['Object ID'], q=4, duplicates='drop')

width_bin_count = met_bins['ID_width_bins'].notna().sum()
quantile_bin_count = met_bins['ID_quantile_bins'].notna().sum()

print('\nRows assigned to width-based bins:', width_bin_count)
print('Rows assigned to quantile-based bins:', quantile_bin_count)

# 4. Work with text fields to extract and standardize information
met_strings = met.copy()

# Convert artist names to uppercase for consistency
met_strings['Artist_Upper'] = met_strings['Artist Display Name'].str.upper()

# Extract the first word from the Medium field
met_strings['Medium_FirstWord'] = met_strings['Medium'].str.split().str[0]

# Identify rows where the Medium description contains the word "wood"
met_strings['Wood_Count'] = met_strings['Medium'].str.contains(
    'wood', case=False, na=False)

artist_upper_count = met_strings['Artist_Upper'].notna().sum()
medium_firstword_count = met_strings['Medium_FirstWord'].notna().sum()
wood_count = met_strings['Wood_Count'].sum()

print('\nArtist names converted to uppercase:', artist_upper_count)
print('Medium first words extracted:', medium_firstword_count)
print('Rows containing the word "wood":', wood_count)

Values mapped for PublicStatus: 484956
AccessionYear values successfully extracted: 481054

Culture values replaced: 141
Department values replaced: 39626

Rows assigned to width-based bins: 484956
Rows assigned to quantile-based bins: 484956

Artist names converted to uppercase: 282513
Medium first words extracted: 477735
Rows containing the word "wood": 35962


<div style="border: 2px solid black; padding: 10px;">

### Summary (Part 2)

This section focused on transforming values, replacing specific entries, creating bins for numeric fields, and working with text fields. These operations helped standardize the dataset, extract useful features, and prepare the data for later grouping and analysis.

**Value Transformations**  
- Public domain values mapped to labels ('Public' / 'Restricted'): **484,956** rows updated  
- AccessionYear values successfully converted to integers: **481,054** rows updated  

These transformations make the fields easier to interpret and more consistent for filtering and classification.

**Value Replacements**  
- Culture values changed from 'Unknown' to 'Not Specified': **141** replacements  
- Department values updated using the mapping dictionary: **39,626** replacements  

Replacing values helps reduce category fragmentation and ensures consistent labeling across the dataset.

**Numeric Binning**  
- Rows assigned to equal‑width bins: **484,956**  
- Rows assigned to quantile‑based bins: **484,956**  

Both binning methods successfully categorized every row, confirming that Object ID contains no missing values and is suitable for numeric segmentation.

**String Processing**  
- Artist names converted to uppercase: **282,513** rows updated  
- First word extracted from Medium field: **477,735** rows updated  
- Rows containing the word “wood”: **35,962** identified  

These string operations help extract meaningful text features and support later text‑based filtering and analysis.

<hr style="border: none; height: 3px; background-color: black;">

### Chapter 8 — Indexing and Merging (Part 1)

In this first part of Chapter 8, I focus on creating a hierarchical index and beginning the process of combining datasets from multiple sources. These steps help standardize the survey data, extract metadata from the PDF, and prepare the 2016 and 2017 datasets for reshaping and merging.

### Steps for this section:

1. Display all unique country variations and apply light standardization to improve consistency before indexing and merging.  
2. Create a hierarchical index using demographic fields from the 2016 survey dataset.  
3. Extract survey question metadata from the 2017 PDF file and convert it into a small DataFrame.  
4. Reshape the 2017 survey responses and merge them with the PDF metadata and the candy reference dataset to create a combined view of the data.

These steps demonstrate how to integrate multiple datasets, handle different formats, and build relationships between files.

<hr style="border: none; height: 3px; background-color: black;">

In [17]:
# 1. Display all unique country variations before standardizing
unique_countries_before = df_2017['Q4: COUNTRY'].unique()
print('Unique country values before cleaning:')
print(unique_countries_before)
print(
    f'\nNumber of unique country variations before cleaning: {len(unique_countries_before)}')

# Light standardization pass: trim, uppercase, and normalize common US variants
df_2016['Which country do you live in?'] = (
    df_2016['Which country do you live in?']
    .astype(str)
    .str.strip()
    .str.upper()
    .replace({'US': 'USA', 'U.S.A': 'USA', 'U.S.A.': 'USA'})
)

df_2017['Q4: COUNTRY'] = (
    df_2017['Q4: COUNTRY']
    .astype(str)
    .str.strip()
    .str.upper()
    .replace({'US': 'USA', 'U.S.A': 'USA', 'U.S.A.': 'USA'})
)

# Show unique values after the initial cleanup
unique_countries_after = df_2017['Q4: COUNTRY'].unique()
print('\nUnique country values after initial cleaning:')
print(unique_countries_after)
print(
    f'\nNumber of unique country variations after initial cleaning: {len(unique_countries_after)}')

# ───────────────────────────────────────────────────────────────
# Full standardization: consolidate US, UK, and Canada variations
# ───────────────────────────────────────────────────────────────
print("\n=== Final Standardization of US, UK, and Canada Variations ===")

# Lists of messy variations to normalize
us_variations = [
    'USA',
    'u.s.',
    'UNITED STATES',
    'UNITED STAES',
    'UNITED STATES OF AMERICA',
    'USAUSAUSA',
    'AMERICA',
    'UNHINGED STATES',
    'US OF A',
    'UNITES STATES',
    'THE UNITED STATES',
    'NORTH CAROLINA',
    'UNIED STATES',
    'U S',
    'THE UNITED STATES OF AMERICA',
    'UNITE STATES',
    'INSANITY LATELY',
    'USA HARD TO TELL ANYMORE…',
    'MERICA',
    'USAS',
    'PITTSBURGH',
    'UNITED STATE',
    'NEW YORK',
    'TRUMPISTAN',
    'UNITED SATES',
    'CALIFORNIA',
    'I PRETEND TO BE FROM CANADA BUT I AM FROM THE UNITED STATES',
    'UNITED STATED',
    'AHEM...AMERCA',
    'NEW JERSEY',
    'UNITED STSTES',
    'UNITED STATSS',
    'MURRIKA',
    'USA! USA! USA!',
    'USAA',
    'ALASKA',
    'USSA',
    'U S A',
    'UNITED STATEA',
    'USA USA USA!!!!',
    'UD',
    'A']

uk_variations = ['UK', 'UNITED KINGDOM', 'ENGLAND', 'U.K.', 'ENDLAND']
canada_variations = ['CANADA', 'CAN', 'CANAE', 'CANADA`', 'CAN CANAE']

# Build mapping dictionaries (lowercased for matching)
us_map = {v.lower().strip(): 'U.S.A.' for v in us_variations}
uk_map = {v.lower().strip(): 'United Kingdom' for v in uk_variations}
canada_map = {v.lower().strip(): 'Canada' for v in canada_variations}

# Combine all mappings
country_map = {**us_map, **uk_map, **canada_map}

# Apply full standardization to both datasets
df_2017['Q4: COUNTRY'] = (
    df_2017['Q4: COUNTRY']
    .astype(str)
    .str.lower()
    .str.strip()
    .replace(country_map)
)

df_2016['Which country do you live in?'] = (
    df_2016['Which country do you live in?']
    .astype(str)
    .str.lower()
    .str.strip()
    .replace(country_map)
)

# Verification counts
print("\nTop 15 country values AFTER full standardization (2017):")
print(df_2017['Q4: COUNTRY'].value_counts().head(15))

usa_count = (df_2017['Q4: COUNTRY'] == 'U.S.A.').sum()
uk_count = (df_2017['Q4: COUNTRY'] == 'United Kingdom').sum()
canada_count = (df_2017['Q4: COUNTRY'] == 'Canada').sum()

print(f"\nTotal standardized as 'U.S.A.': {usa_count}")
print(f"Total standardized as 'United Kingdom': {uk_count}")
print(f"Total standardized as 'Canada': {canada_count}")

final_unique = df_2017['Q4: COUNTRY'].nunique()
print(
    f"Final unique country variations after US + UK + Canada cleanup: {final_unique}")

# 2. Create a hierarchical index using demographic fields from the 2016 survey
df_2016_indexed = df_2016.set_index([
    'Which country do you live in?',
    'Which state, province, county do you live in?',
    'Your gender:',
    'How old are you?'
])

print('\nHierarchical index created. Index levels:')
print(df_2016_indexed.index.names)

# 3. Extract survey question metadata from the 2017 PDF (manually
# reconstructed)
pdf_metadata = {
    'Question': ['Q1', 'Q2', 'Q3', 'Q4', 'Q5'],
    'Prompt': [
        'Are you actually going trick or treating yourself?',
        'Your gender',
        'How old are you?',
        'What country do you live in?',
        'Which state, province, county do you live in?'
    ]
}
df_pdf_meta = pd.DataFrame(pdf_metadata)
print('\nPDF metadata table created:')
print(df_pdf_meta)

# 4. Reshape the 2017 survey and merge with metadata and candy reference data

# Identify ID fields and candy question columns
id_vars = [
    'Internal ID',
    'Q1: GOING OUT?',
    'Q2: GENDER',
    'Q3: AGE',
    'Q4: COUNTRY',
    'Q5: STATE, PROVINCE, COUNTY, ETC'
]
value_vars = [col for col in df_2017.columns if col.startswith('Q6')]

# Melt the dataset from wide to long format
original_rows = len(df_2017)
df_2017_long = df_2017.melt(
    id_vars=id_vars,
    value_vars=value_vars,
    var_name='Question_Candy',
    value_name='Response'
)

melted_rows = len(df_2017_long)
print(f'\nOriginal 2017 rows: {original_rows}')
print(f'Rows after melting: {melted_rows}')
print(f'Rows created during melt: {melted_rows - original_rows}')

# Clean candy names for merging
df_2017_long['Candy'] = df_2017_long['Question_Candy'].str.replace(
    'Q6 \\| ', '', regex=True)

# Merge with candy metadata
df_merged = df_2017_long.merge(
    df_candydata,
    left_on='Candy',
    right_on='ITEM',
    how='left'
)

merged_rows = len(df_merged)
unmatched = df_merged['JOY'].isna().sum()

print(f'\nRows after merge: {merged_rows}')
print(f'Unmatched candy items (no metadata found): {unmatched}')
print('\nMerged dataset preview:')
print(df_merged.head())

Unique country values before cleaning:
[nan 'USA ' 'USA' 'us' 'usa' 'canada' 'Canada' 'Us' 'US' 'Murica'
 'United States' 'uk' 'United Kingdom' 'united states' 'Usa'
 'United States ' 'United staes' 'United States of America' 'UAE'
 'England' 'UK' 'canada ' 'Mexico' 'United states' 'u.s.a.' 'USAUSAUSA'
 'america' 35 'france' 'United States of America ' 'U.S.A.' 'finland'
 'unhinged states' 'Canada ' 'united states of america' 'US of A'
 'Unites States' 'The United States' 'North Carolina ' 'Unied States'
 'Netherlands' 'germany' 'Europe' 'Earth' 'U S' 'u.s.' 'U.K. '
 'Costa Rica' 'The United States of America' 'unite states' 'U.S.' 46
 'cascadia' 'Australia' 'insanity lately' 'Greece'
 'USA? Hard to tell anymore..' "'merica" 'usas' 'Pittsburgh' 45
 'United State' 32 'France' 'australia' 'A' 'Can' 'Canae' 'New York'
 'Trumpistan' 'Ireland' 'United Sates' 'Korea' 'California' 'Japan' 'USa'
 'South africa'
 'I pretend to be from Canada, but I am really from the United States.'
 'Usa ' 'Uk

<div style="border: 2px solid black; padding: 10px;">

### Summary (Part 1)

Before cleaning, the 2017 dataset contained **129** unique country variations. After the initial standardization pass, **94** unique values remained. Many of these residual entries were still variations of the United States, often due to misspellings, jokes, fictional places, or responses that referenced states, continents, or non‑country locations. This confirmed the need for a more thorough standardization process.

In this part of Chapter 8, I:

- Applied country standardization by:
    - Trimming whitespace  
    - Converting values to uppercase  
    - Consolidating common variants such as “US” and “U.S.A.” into “USA” during the initial cleaning  
- Created a hierarchical index for the 2016 survey using four demographic fields to support multi‑level analysis.  
- Extracted the key survey prompts (Q1–Q5) from the 2017 PDF and converted them into a small metadata table for merging.  
- Reshaped the 2017 survey from wide to long format, expanding the dataset from **2,460** rows to **253,380** rows and generating **250,920** new rows.  
- Cleaned the candy names and merged the reshaped data with the candy reference dataset, producing **253,380** merged rows.  
- Identified **81,180** candy entries that did not match the reference file, which is expected due to joke responses, novelty items, non‑candy entries, and intentionally inconsistent survey answers.

<hr style="border: none; height: 3px; background-color: black;">

### Chapter 8 — Part 2  
Reshaping and Pivoting the Data

In this second part of Chapter 8, I focus on reshaping the merged dataset by creating pivot tables and reorganizing the data into more analytical forms. These steps help summarize candy sentiment, compare preferences across demographic groups, and examine response distributions.

### Steps for this section:

1. Create a pivot table summarizing JOY, DESPAIR, and NET FEELIES by candy item.  
2. Create a pivot table summarizing JOY, DESPAIR, and NET FEELIES by country.  
3. Reshape the merged dataset to compare candy preferences across demographic fields.  
4. Generate a pivot table showing the distribution of JOY/MEH/DESPAIR responses for each candy.

<hr style="border: none; height: 3px; background-color: black;">

In [18]:
# 1. Pivot JOY, DESPAIR, and NET FEELIES by candy item
# Summarizes average sentiment scores for each candy across all responses
pivot_candy = df_merged.pivot_table(
    index='ITEM',
    values=['JOY', 'DESPAIR', 'NET FEELIES'],
    aggfunc='mean'
)

print('Pivot table by candy item:')
print(pivot_candy.head())


# 2. Pivot JOY, DESPAIR, and NET FEELIES by country
# Shows how sentiment varies across countries in the survey
pivot_country = df_merged.pivot_table(
    index='Q4: COUNTRY',
    values=['JOY', 'DESPAIR', 'NET FEELIES'],
    aggfunc='mean'
)

print('\nPivot table by country:')
print(pivot_country.head())


# 3. Reshape merged dataset to compare candy preferences across demographics
# Creates a multi-index pivot comparing JOY scores by gender and age
pivot_demographics = df_merged.pivot_table(
    index=['Q2: GENDER', 'Q3: AGE'],
    columns='ITEM',
    values='JOY',
    aggfunc='mean'
)

print('\nPivot table by gender and age:')
print(pivot_demographics.head())


# 4. Pivot table showing JOY/MEH/DESPAIR distribution for each candy
# Counts how many times each candy received each type of response
pivot_responses = df_merged.pivot_table(
    index='ITEM',
    columns='Response',
    values='Internal ID',
    aggfunc='count',
    fill_value=0
)

print('\nPivot table of JOY/MEH/DESPAIR counts by candy:')
# 1. Pivot JOY, DESPAIR, and NET FEELIES by candy item
print(pivot_responses.head())
# Summarizes average sentiment scores for each candy across all responses
pivot_candy = df_merged.pivot_table(
    index='ITEM',
    values=['JOY', 'DESPAIR', 'NET FEELIES'],
    aggfunc='mean'
)

print('Pivot table by candy item:')
print(pivot_candy.head())


# 2. Pivot JOY, DESPAIR, and NET FEELIES by country
# Shows how sentiment varies across countries in the survey
pivot_country = df_merged.pivot_table(
    index='Q4: COUNTRY',
    values=['JOY', 'DESPAIR', 'NET FEELIES'],
    aggfunc='mean'
)

print('\nPivot table by country:')
print(pivot_country.head())


# 3. Reshape merged dataset to compare candy preferences across demographics
# Creates a multi-index pivot comparing JOY scores by gender and age
pivot_demographics = df_merged.pivot_table(
    index=['Q2: GENDER', 'Q3: AGE'],
    columns='ITEM',
    values='JOY',
    aggfunc='mean'
)

print('\nPivot table by gender and age:')
print(pivot_demographics.head())


# 4. Pivot table showing JOY/MEH/DESPAIR distribution for each candy
# Counts how many times each candy received each type of response
pivot_responses = df_merged.pivot_table(
    index='ITEM',
    columns='Response',
    values='Internal ID',
    aggfunc='count',
    fill_value=0
)

print('\nPivot table of JOY/MEH/DESPAIR counts by candy:')
print(pivot_responses.head())

Pivot table by candy item:
                          DESPAIR    JOY  NET FEELIES
ITEM                                                 
100 Grand Bar                34.0  570.0        536.0
Any full-sized candy bar     13.0  828.0        815.0
Black Jacks                 252.0   34.0       -218.0
Bottle Caps                 219.0  300.0         81.0
Broken glow stick           579.0   14.0       -565.0

Pivot table by country:
                DESPAIR         JOY  NET FEELIES
Q4: COUNTRY                                     
'merica      194.114286  334.414286        140.3
1            194.114286  334.414286        140.3
32           194.114286  334.414286        140.3
35           194.114286  334.414286        140.3
45           194.114286  334.414286        140.3

Pivot table by gender and age:
ITEM                100 Grand Bar  Any full-sized candy bar  Black Jacks  \
Q2: GENDER Q3: AGE                                                         
Female     6                570.0          

<div style="border: 2px solid black; padding: 10px;">

### Summary (Part 2)

In this part of Chapter 8, I reshaped the merged dataset and created several pivot tables to summarize candy sentiment, compare demographic patterns, and analyze response distributions.

- Created a pivot table summarizing **JOY, DESPAIR, and NET FEELIES** for each candy item. This highlighted candies with strong positive sentiment (such as full‑sized candy bars, Kit Kat, and Reese’s Peanut Butter Cups) as well as candies with strongly negative sentiment (such as Broken glow stick, Black Jacks, and Creepy Religious Comics). These values come directly from the candy reference dataset and provide a clear ranking of overall candy appeal.

- Generated a pivot table summarizing **JOY, DESPAIR, and NET FEELIES** by country. Because the country field contains many humorous, fictional, or inconsistent entries, several countries share identical aggregated values. This reflects the intentionally chaotic nature of the survey responses rather than meaningful geographic differences.

- Created a demographic pivot table comparing candy preferences by **gender and age**. The pivot table shows consistent JOY values across age groups for each candy, which is expected because the JOY scores originate from the candy reference dataset rather than individual respondent ratings. This demonstrates how pivot tables can be used to compare sentiment across demographic categories even when the underlying values are constant.

- Generated a pivot table showing the distribution of **JOY, MEH, and DESPAIR** responses for each candy. This table provides a clearer view of how respondents actually reacted to each candy item. For example, *100 Grand Bar* received **873** JOY responses, **755** MEH responses, and **85** DESPAIR responses, while *Broken glow stick* received **1,625** DESPAIR responses and only **24** JOY responses. These counts reveal the real‑world sentiment patterns behind the candy rankings.

<hr style="border: none; height: 3px; background-color: black;">

### Chapter 10 — Grouping and Split/Apply/Combine

In this chapter, I focus on grouping operations using dictionaries, Series objects, custom functions, and index levels, as well as applying the split–apply–combine pattern to the reshaped candy survey dataset. These techniques make it possible to summarize sentiment, compare demographic patterns, and analyze response distributions across multiple dimensions.

### Steps:

1. Use dictionaries and Series objects to group candy items into broader categories (e.g., chocolate, fruit, novelty, non‑candy) and compute aggregated JOY, MEH, and DESPAIR values for each group.  
2. Apply custom functions to group respondents into age brackets and analyze how candy sentiment varies across age groups.  
3. Use index‑level grouping to summarize JOY, DESPAIR, and NET FEELIES using the hierarchical index created earlier in Chapter 8.  
4. Demonstrate the split–apply–combine pattern by computing grouped statistics across demographic fields such as gender, age, and country.  
5. Generate cross‑tabulations (crosstabs) to compare JOY/MEH/DESPAIR distributions across demographic categories and candy types.

<hr style="border: none; height: 3px; background-color: black;">

In [19]:
# 1. Grouping with dictionaries and Series
# Map each candy item to a broader category (e.g., Chocolate, Fruit Candy)
candy_categories = {
    '100 Grand Bar': 'Chocolate',
    'Any full-sized candy bar': 'Chocolate',
    'Black Jacks': 'Licorice',
    'Bottle Caps': 'Sugar Candy',
    'Broken glow stick': 'Non-Candy',
    'Candy Corn': 'Sugar Candy',
    'Kit Kat': 'Chocolate',
    'Reese’s Peanut Butter Cups': 'Chocolate',
    'Skittles': 'Fruit Candy',
    'Starburst': 'Fruit Candy',
    'Swedish Fish': 'Fruit Candy'
}

category_series = pd.Series(candy_categories, name='Category')

# Add category labels to the merged dataset
df_merged['Category'] = df_merged['ITEM'].map(category_series)

# Compute average sentiment scores for each candy category
grouped_categories = df_merged.groupby(
    'Category')[['JOY', 'DESPAIR', 'NET FEELIES']].mean()

print('Grouped by candy category:')
print(grouped_categories.head())


# 2. Grouping with a custom function (age brackets)
# Convert raw ages into bracketed age groups
def age_bracket(age):
    try:
        age = int(age)
    except BaseException:
        return 'Unknown'
    if age < 18:
        return 'Under 18'
    elif age < 30:
        return '18–29'
    elif age < 50:
        return '30–49'
    else:
        return '50+'


df_merged['Age Bracket'] = df_merged['Q3: AGE'].apply(age_bracket)

# Compute average sentiment scores by age bracket
grouped_age = df_merged.groupby('Age Bracket')[
    ['JOY', 'DESPAIR', 'NET FEELIES']].mean()

print('\nGrouped by age bracket:')
print(grouped_age)


# 3. Grouping with index levels (gender + age)
# Use a multi-index to group by gender and age simultaneously
df_indexed = df_merged.set_index(['Q2: GENDER', 'Q3: AGE'])

grouped_index_levels = df_indexed.groupby(
    level=[0, 1])[['JOY', 'DESPAIR']].mean()

print('\nGrouped by gender and age (index levels):')
print(grouped_index_levels.head())


# 4. Split–apply–combine (mean JOY per candy per gender)
# Demonstrates the classic pattern: split data → apply function → combine
# results
split_apply = df_merged.groupby(['Q2: GENDER', 'ITEM'])['JOY'].mean()

print('\nSplit–apply–combine (JOY by gender and candy):')
print(split_apply.head())


# 5. Crosstab of JOY/MEH/DESPAIR by gender
# Shows the distribution of response types for each gender
crosstab_responses = pd.crosstab(
    df_merged['Q2: GENDER'],
    df_merged['Response'],
    normalize='index'
)

print('\nCrosstab of response distribution by gender:')
print(crosstab_responses)

Grouped by candy category:
                    JOY  DESPAIR  NET FEELIES
Category                                     
Chocolate    818.750000    27.25   791.500000
Fruit Candy  493.666667   109.00   384.666667
Licorice      34.000000   252.00  -218.000000
Non-Candy     14.000000   579.00  -565.000000
Sugar Candy  378.500000   273.00   105.500000

Grouped by age bracket:
                    JOY     DESPAIR  NET FEELIES
Age Bracket                                     
18–29        334.414286  194.114286        140.3
30–49        334.414286  194.114286        140.3
50+          334.414286  194.114286        140.3
Under 18     334.414286  194.114286        140.3
Unknown      334.414286  194.114286        140.3

Grouped by gender and age (index levels):
                           JOY     DESPAIR
Q2: GENDER Q3: AGE                        
Female     6        334.414286  194.114286
           7        334.414286  194.114286
           8        334.414286  194.114286
           9        334.4

<div style="border: 2px solid black; padding: 10px;">

### Summary (Chapter 10)

In this chapter, I applied grouping techniques and the split–apply–combine pattern to the reshaped candy survey dataset. These operations helped summarize sentiment across candy categories, demographic groups, and response types.

- Grouped candies into broader categories (Chocolate, Fruit Candy, Sugar Candy, Licorice, and Non‑Candy) using a dictionary and computed mean `JOY`, `DESPAIR`, and `NET FEELIES` for each group. Chocolate items showed the highest average `JOY` (**818.75**), while Non‑Candy items showed the strongest negative sentiment (`DESPAIR` = **579**, `NET FEELIES` = **–565**).

- Created age brackets using a custom function and grouped the dataset by these brackets. Because `JOY`, `DESPAIR`, and `NET FEELIES` originate from the candy reference dataset, all age brackets produced identical averages. This illustrates how grouping functions behave even when the underlying values are constant.

- Used index‑level grouping with the hierarchical index (gender + age) created in Chapter 8. This produced consistent `JOY` and `DESPAIR` values across demographic combinations, reinforcing how multi‑level indexes can be used to organize and summarize data.

- Demonstrated the split–apply–combine pattern by computing mean `JOY` values for each candy within each gender category. For example, Female respondents showed high `JOY` for items like *Any full‑sized candy bar* (828) and *Kit Kat* (894), and very low `JOY` for *Broken glow stick* (14).

- Generated a crosstab comparing `JOY`, `MEH`, and `DESPAIR` distributions across gender categories. The results show similar sentiment patterns across groups, with `JOY` making up roughly **36–38%** of responses, `MEH` around **28–32%**, and `DESPAIR` around **32–33%**.

<hr style="border: none; height: 3px; background-color: black;">

### Chapter 11 — Working with Dates and Times

In this chapter, I demonstrate how to convert string fields into datetime objects, generate date ranges, work with frequencies and date offsets, and convert between timestamps and periods. These operations form the foundation of time‑series analysis and allow for more flexible indexing, grouping, and resampling.

### Steps for this section:

1. Convert year columns to numeric values to ensure they can be used in datetime operations.  
2. Create an annual `AccessionPeriod` to represent each year as a period rather than a timestamp.  
3. Apply date offsets such as `MonthEnd`, `YearBegin`, and `BusinessDay` to explore how timestamps shift under different offset rules.  
4. Convert timestamps to periods (quarter and decade) to demonstrate how pandas handles coarser time frequencies.  
5. Summarize the results and verify that the period conversions behave as expected.

<hr style="border: none; height: 3px; background-color: black;">

In [ ]:
# 1. Convert year columns to numeric
# Convert year-like strings into integers using a strict YYYY format
met['AccessionYear'] = pd.to_datetime(
    met['AccessionYear'], format='%Y', errors='coerce'
).dt.year.astype('Int64')

met['Object Begin Date'] = pd.to_datetime(
    met['Object Begin Date'], format='%Y', errors='coerce'
).dt.year.astype('Int64')

met['Object End Date'] = pd.to_datetime(
    met['Object End Date'], format='%Y', errors='coerce'
).dt.year.astype('Int64')

print('Converted year columns:')
print(met[['AccessionYear', 'Object Begin Date', 'Object End Date']].head())


# 2. Create an annual AccessionPeriod
# Reset index to avoid alignment issues when updating PeriodIndex
met = met.reset_index(drop=True)

# Extract valid years as strings for PeriodIndex construction
valid_years_str = met['AccessionYear'].dropna().astype(str)

# Create a PeriodIndex anchored to year-end (Dec 31)
period_index = pd.PeriodIndex(valid_years_str, freq='Y-DEC')

# Initialize a full-length period Series and update only valid rows
accession_period = pd.Series(pd.NaT, index=met.index, dtype='period[Y]')
accession_period.update(
    pd.Series(period_index, index=met['AccessionYear'].dropna().index)
)

met['AccessionPeriod'] = accession_period

# Convert periods back to timestamps for display
met['AccessionTimestamp'] = met['AccessionPeriod'].dt.to_timestamp(how='end')
met['AccessionTimestamp_Display'] = met['AccessionTimestamp'].dt.strftime(
    '%Y-%m-%d')

print('\nAccessionPeriod sample:')
print(met[['AccessionYear', 'AccessionPeriod',
      'AccessionTimestamp_Display']].head())


# 3. Apply date offsets
# Demonstrate MonthEnd, YearBegin, and BusinessDay offsets
example_date = pd.Timestamp('2020-01-15')

shift_month_end = example_date + pd.offsets.MonthEnd(1)
shift_year_begin = example_date + pd.offsets.YearBegin(1)
shift_business_day = example_date + pd.offsets.BusinessDay(5)

print('\nDate offsets:')
print('MonthEnd:', shift_month_end)
print('YearBegin:', shift_year_begin)
print('BusinessDay:', shift_business_day)


# 4. Convert timestamps to periods (quarter and decade)
# Convert Dec 31 timestamps to quarterly periods
met['AccessionQuarter'] = met['AccessionTimestamp'].dt.to_period('Q')

# Convert timestamps to decade periods using integer math
met['AccessionDecade'] = met['AccessionTimestamp'].apply(
    lambda ts: pd.Period(
        year=(ts.year // 10) * 10,
        freq='10Y'
    ) if pd.notna(ts) else pd.NaT
)

print('\nQuarter and Decade sample:')
print(met[['AccessionYear', 'AccessionPeriod',
           'AccessionQuarter', 'AccessionDecade']].head())


# 5. Summary
print('\nSummary:')
print('Total rows:', len(met))
print('Unique Accession Years:', met['AccessionYear'].dropna().nunique())
print('Earliest year:', met['AccessionYear'].min())
print('Latest year:', met['AccessionYear'].max())
print('Unique Quarters:', met['AccessionQuarter'].dropna().nunique())
print('Unique Decades:', met['AccessionDecade'].dropna().nunique())

Converted year columns:
   AccessionYear  Object Begin Date  Object End Date
0           1979               1853             1853
1           1980               1901             1901
2           1967               1909             1927
3           1967               1909             1927
4           1967               1909             1927


<div style="border: 2px solid black; padding: 10px;">

### Summary

In this section, I worked with year fields, timestamps, and period-based representations to explore how pandas handles date and time data:

- Converted the year columns into numeric `Int64` values to ensure they could be used in datetime and period operations.  
- Created an annual `AccessionPeriod` anchored to year‑end and generated corresponding timestamps for display.  
- Applied several date offsets (`MonthEnd`, `YearBegin`, `BusinessDay`) to demonstrate how timestamps shift under different offset rules.  
- Converted timestamps to quarterly and decade‑level periods, showing how pandas supports coarser time frequencies.  
- Added summary statistics to confirm the range of years, quarters, and decades present in the dataset and to verify that the period conversions behaved as expected.

<div style="border: 2px solid black; padding: 12px;">

### Reflection — Week 7 & 8

### Overall Assessment  
The assignments definitely provided a lot of real world practice through a wide range of data transformation techniques that covered cleaning, reshaping, merging, grouping, and time‑series manipulation. Each chapter built on the previous one, moving from foundational transformations (mapping, replacing, binning, string operations) to more advanced reshaping (melt, pivot, hierarchical indexing), and finally into grouping logic and date/period conversions. Together, these chapters reinforced how flexible pandas is when working with messy, multi‑source datasets and how important it is to understand the structure of the data before choosing the right transformation. Overall, this sequence strengthened my confidence in preparing datasets for analysis and in applying transformations that make the data more meaningful and easier to work with.

---

### Easier Portions  
These tasks felt intuitive and aligned well with earlier work:

- **Mapping, replacing, and string transformations:**  
  Using `.map()`, `.replace()`, and `.str` methods felt natural and produced clear, immediate improvements in data consistency.

- **Initial reshaping and pivoting:**  
  Creating pivot tables and reorganizing the data with `pivot_table()` was straightforward once the dataset was in long format.

- **Grouping with dictionaries and Series:**  
  Assigning candy categories and computing aggregated sentiment values was a clean, predictable workflow.

- **Basic datetime conversions:**  
  Converting year fields with `to_datetime()` and extracting `.dt.year` was simple and easy to validate.

---

### More Challenging Portions  
These steps required more careful reasoning and attention to detail:

- **Cleaning and standardizing country names:**  
  The large number of humorous, fictional, and inconsistent entries made it difficult to create a clean mapping. Ensuring that the mapping logic didn’t overwrite valid values required deliberate planning.

- **Melting and merging the 2017 survey:**  
  Reshaping the dataset from wide to long format and merging with the candy reference file introduced alignment issues and required careful column selection.

- **Multi‑level grouping and hierarchical indexing:**  
  Working with index levels and grouping across gender + age combinations required a deeper understanding of how pandas handles multi‑index structures.

- **Period and timestamp conversions:**  
  Understanding how periods anchor to timestamps, how offsets shift dates, and how decade‑level periods are constructed required more conceptual reasoning than earlier chapters.

---

### Start / Stop / Continue

**Start**  
- Validating transformations with small samples before applying them to the full dataset.  
- Using more visual summaries (e.g., value counts, histograms) to confirm that transformations behave as expected.  
- Incorporating more defensive programming when working with messy fields (e.g., country names, ages).

**Stop**  
- Assuming that similar‑looking fields (like year columns) are clean or consistently formatted.  
- Treating reshaped datasets as error‑free — melting and merging often introduce unexpected alignment issues.  
- Relying on implicit index behavior; explicit resets and alignment checks prevent subtle bugs.

**Continue**  
- Using mapping, replacing, and string operations as first‑line tools for cleaning categorical fields.  
- Applying pivot tables and grouping operations to explore patterns across categories and demographics.  
- Leveraging period and timestamp conversions when working with time‑based data, especially when analyzing trends across years, quarters, or decades.  
- Writing clear, intentional comments that explain the purpose behind each transformation.
